File: engine/incidents.ts
export type IncidentType =
  | 'ddos'
  | 'disk_full'
  | 'vendor_offer'
  | 'memory_leak'
  | 'hacker_breach';

export type ActiveIncident =
  | {
      type: 'ddos';
      startedAt: number;
      expiresAt: number;
      tapsRemaining: number;
    }
  | { type: 'disk_full'; startedAt: number; expiresAt: number }
  | { type: 'vendor_offer'; startedAt: number; expiresAt: number }
  | { type: 'memory_leak'; startedAt: number; expiresAt: number }
  | {
      type: 'hacker_breach';
      startedAt: number;
      expiresAt: number;
      sequence: string[]; // letters in the order to tap
      buttons: string[]; // letters as displayed (shuffled)
      currentStep: number; // index into sequence
    };

export interface IncidentResolution {
  success: boolean;
  rewardCredits: number;
  penaltyCredits: number;
  vendorDiscountActivated: boolean;
}

// ─── Tunable config per incident type ───
export const INCIDENT_CONFIG = {
  ddos: {
    name: 'DDoS Attack',
    icon: '🛑',
    description: 'Inbound flood is choking your servers. Mitigate fast.',
    duration: 15, // seconds (was 30)
    tapsRequired: 3,
    activeMultiplier: 0.5, // output while incident is live
    rewardSecondsOfCps: 30,
    timeoutPenaltySecondsOfCps: 120, // harsher (was 60)
  },
  disk_full: {
    name: 'Disk Full',
    icon: '💾',
    description: 'A runaway log filled the disk. Output is throttled.',
    duration: 20, // seconds (was 45)
    activeMultiplier: 0.5,
    rewardSecondsOfCps: 15,
    timeoutPenaltySecondsOfCps: 30, // NEW — was silent
  },
  vendor_offer: {
    name: 'Vendor Offer',
    icon: '💰',
    description: 'Sales rep is offering 50% off your next server purchase.',
    duration: 30, // seconds (was 60)
    activeMultiplier: 1, // no penalty
  },
  memory_leak: {
    name: 'Memory Leak',
    icon: '🧠',
    description: 'A buggy service is leaking RAM. Output is degrading.',
    duration: 30, // expires after 30s (was 60)
    decayDuration: 15, // takes 15s to decay from 1.0 to 0.5 (was 30)
    minMultiplier: 0.25, // floor below the 0.5 mark
    activeMultiplier: 1, // base, but recomputed dynamically
    rewardSecondsOfCps: 20,
    timeoutCreditPercent: 0.02, // NEW — lose 2% credits on timeout
  },
  hacker_breach: {
    name: 'Hacker Breach',
    icon: '🚨',
    description: 'Someone is in. Patch the firewall in order.',
    duration: 12, // 12 seconds to complete the sequence
    sequenceLength: 4,
    activeMultiplier: 0.7, // mild output penalty while active
    rewardSecondsOfCps: 90,
    timeoutCreditPercent: 0.10, // harsher (was 0.05)
  },
} as const;

export const INCIDENT_TRIGGER_CHANCE_PER_SEC = 0.01; // ~1% per tick

const ALL_TYPES: IncidentType[] = [
  'ddos',
  'disk_full',
  'vendor_offer',
  'memory_leak',
  'hacker_breach',
];

export function pickRandomIncident(
  weights?: Partial<Record<IncidentType, number>>
): IncidentType | null {
  const w: Record<IncidentType, number> = {
    ddos: weights?.ddos ?? 1,
    disk_full: weights?.disk_full ?? 1,
    vendor_offer: weights?.vendor_offer ?? 1,
    memory_leak: weights?.memory_leak ?? 1,
    hacker_breach: weights?.hacker_breach ?? 1,
  };
  const total = ALL_TYPES.reduce((s, t) => s + w[t], 0);
  if (total <= 0) return null; // fully blocked by staff
  let roll = Math.random() * total;
  for (const t of ALL_TYPES) {
    roll -= w[t];
    if (roll <= 0) return t;
  }
  return ALL_TYPES[0];
}

const HACKER_LETTER_POOL = ['A', 'B', 'C', 'D', 'E', 'F', 'G', 'H'];

function generateHackerSequence(length: number): {
  sequence: string[];
  buttons: string[];
} {
  // Pick `length` unique letters from the pool
  const shuffled = [...HACKER_LETTER_POOL].sort(() => Math.random() - 0.5);
  const letters = shuffled.slice(0, length);
  // The sequence is the letters in a random order (must be tapped in this order)
  const sequence = [...letters].sort(() => Math.random() - 0.5);
  // Buttons display the same letters in another random order
  const buttons = [...letters].sort(() => Math.random() - 0.5);
  return { sequence, buttons };
}

export function createIncident(type: IncidentType): ActiveIncident {
  const now = Date.now();
  const config = INCIDENT_CONFIG[type];
  const expiresAt = now + config.duration * 1000;

  switch (type) {
    case 'ddos':
      return {
        type,
        startedAt: now,
        expiresAt,
        tapsRemaining: INCIDENT_CONFIG.ddos.tapsRequired,
      };
    case 'disk_full':
    case 'vendor_offer':
    case 'memory_leak':
      return { type, startedAt: now, expiresAt };
    case 'hacker_breach': {
      const { sequence, buttons } = generateHackerSequence(
        INCIDENT_CONFIG.hacker_breach.sequenceLength
      );
      return {
        type,
        startedAt: now,
        expiresAt,
        sequence,
        buttons,
        currentStep: 0,
      };
    }
  }
}

export function getIncidentMultiplier(incident: ActiveIncident | null): number {
  if (!incident) return 1;
  if (incident.type === 'memory_leak') {
    // Decays linearly from 1.0 → 0.5 over decayDuration, then floors
    const config = INCIDENT_CONFIG.memory_leak;
    const elapsedSec = (Date.now() - incident.startedAt) / 1000;
    const decayProgress = Math.min(1, elapsedSec / config.decayDuration);
    const value = 1 - decayProgress * 0.5;
    return Math.max(config.minMultiplier, value);
  }
  return INCIDENT_CONFIG[incident.type].activeMultiplier;
}

export function getIncidentTimeRemaining(
  incident: ActiveIncident | null
): number {
  if (!incident) return 0;
  return Math.max(0, (incident.expiresAt - Date.now()) / 1000);
}

File: lib/engine/incidents.dart
import 'dart:math';

// ─── Incident types ───

enum IncidentType {
  ddos,
  diskFull,
  memoryLeak,
  vendorOffer,
  hackerBreach,
}

// ─── Per-type config ───

class IncidentConfig {
  final String name;
  final String icon;
  final String description;
  final int duration; // seconds
  final int tapsRequired;
  final double activeMultiplier;
  final double rewardSecondsOfCps;
  final double timeoutPenaltySecondsOfCps;
  final double timeoutCreditPercent;
  final int sequenceLength;
  final int decayDuration;
  final double minMultiplier;

  const IncidentConfig({
    required this.name,
    required this.icon,
    required this.description,
    required this.duration,
    this.tapsRequired = 0,
    this.activeMultiplier = 1.0,
    this.rewardSecondsOfCps = 0,
    this.timeoutPenaltySecondsOfCps = 0,
    this.timeoutCreditPercent = 0,
    this.sequenceLength = 0,
    this.decayDuration = 0,
    this.minMultiplier = 0,
  });
}

const Map<IncidentType, IncidentConfig> incidentConfigs = {
  IncidentType.ddos: IncidentConfig(
    name: 'DDoS Attack',
    icon: '\u{1F6D1}', // red octagonal
    description: 'Inbound flood is choking your servers. Mitigate fast.',
    duration: 15,
    tapsRequired: 3,
    activeMultiplier: 0.5,
    rewardSecondsOfCps: 30,
    timeoutPenaltySecondsOfCps: 120,
  ),
  IncidentType.diskFull: IncidentConfig(
    name: 'Disk Full',
    icon: '\u{1F4BE}', // floppy disk
    description: 'A runaway log filled the disk. Output is throttled.',
    duration: 20,
    activeMultiplier: 0.5,
    rewardSecondsOfCps: 15,
    timeoutPenaltySecondsOfCps: 30,
  ),
  IncidentType.vendorOffer: IncidentConfig(
    name: 'Vendor Offer',
    icon: '\u{1F4B0}', // money bag
    description: 'Sales rep is offering 50% off your next server purchase.',
    duration: 30,
    activeMultiplier: 1,
  ),
  IncidentType.memoryLeak: IncidentConfig(
    name: 'Memory Leak',
    icon: '\u{1F9E0}', // brain
    description: 'A buggy service is leaking RAM. Output is degrading.',
    duration: 30,
    decayDuration: 15,
    minMultiplier: 0.25,
    activeMultiplier: 1,
    rewardSecondsOfCps: 20,
    timeoutCreditPercent: 0.02,
  ),
  IncidentType.hackerBreach: IncidentConfig(
    name: 'Hacker Breach',
    icon: '\u{1F6A8}', // rotating light
    description: 'Someone is in. Patch the firewall in order.',
    duration: 12,
    sequenceLength: 4,
    activeMultiplier: 0.7,
    rewardSecondsOfCps: 90,
    timeoutCreditPercent: 0.10,
  ),
};

const double incidentTriggerChancePerSec = 0.01;

// ─── Active incident (single class with nullable fields) ───

class ActiveIncident {
  final IncidentType type;
  final int startedAt;
  final int expiresAt;
  final int? tapsRemaining;
  final List<String>? sequence;
  final List<String>? buttons;
  final int? currentStep;

  const ActiveIncident({
    required this.type,
    required this.startedAt,
    required this.expiresAt,
    this.tapsRemaining,
    this.sequence,
    this.buttons,
    this.currentStep,
  });

  ActiveIncident copyWith({
    IncidentType? type,
    int? startedAt,
    int? expiresAt,
    int? tapsRemaining,
    List<String>? sequence,
    List<String>? buttons,
    int? currentStep,
  }) {
    return ActiveIncident(
      type: type ?? this.type,
      startedAt: startedAt ?? this.startedAt,
      expiresAt: expiresAt ?? this.expiresAt,
      tapsRemaining: tapsRemaining ?? this.tapsRemaining,
      sequence: sequence ?? this.sequence,
      buttons: buttons ?? this.buttons,
      currentStep: currentStep ?? this.currentStep,
    );
  }
}

// ─── Weighted random pick ───

const List<IncidentType> _allTypes = [
  IncidentType.ddos,
  IncidentType.diskFull,
  IncidentType.vendorOffer,
  IncidentType.memoryLeak,
  IncidentType.hackerBreach,
];

IncidentType? pickRandomIncident(
  Map<IncidentType, double> weights,
  Random rng,
) {
  final Map<IncidentType, double> w = {
    IncidentType.ddos: weights[IncidentType.ddos] ?? 1,
    IncidentType.diskFull: weights[IncidentType.diskFull] ?? 1,
    IncidentType.vendorOffer: weights[IncidentType.vendorOffer] ?? 1,
    IncidentType.memoryLeak: weights[IncidentType.memoryLeak] ?? 1,
    IncidentType.hackerBreach: weights[IncidentType.hackerBreach] ?? 1,
  };

  final total = _allTypes.fold<double>(0, (s, t) => s + w[t]!);
  if (total <= 0) return null; // fully blocked by staff

  double roll = rng.nextDouble() * total;
  for (final t in _allTypes) {
    roll -= w[t]!;
    if (roll <= 0) return t;
  }
  return _allTypes[0];
}

// ─── Hacker sequence generator ───

const List<String> _hackerLetterPool = [
  'A', 'B', 'C', 'D', 'E', 'F', 'G', 'H',
];

Map<String, List<String>> _generateHackerSequence(int length, Random rng) {
  final shuffled = List<String>.from(_hackerLetterPool)..shuffle(rng);
  final letters = shuffled.sublist(0, length);
  final sequence = List<String>.from(letters)..shuffle(rng);
  final buttons = List<String>.from(letters)..shuffle(rng);
  return {'sequence': sequence, 'buttons': buttons};
}

// ─── Create incident ───

ActiveIncident createIncident(IncidentType type, [Random? rng]) {
  final now = DateTime.now().millisecondsSinceEpoch;
  final config = incidentConfigs[type]!;
  final expiresAt = now + config.duration * 1000;
  final random = rng ?? Random();

  switch (type) {
    case IncidentType.ddos:
      return ActiveIncident(
        type: type,
        startedAt: now,
        expiresAt: expiresAt,
        tapsRemaining: incidentConfigs[IncidentType.ddos]!.tapsRequired,
      );
    case IncidentType.diskFull:
    case IncidentType.vendorOffer:
    case IncidentType.memoryLeak:
      return ActiveIncident(
        type: type,
        startedAt: now,
        expiresAt: expiresAt,
      );
    case IncidentType.hackerBreach:
      final result = _generateHackerSequence(
        incidentConfigs[IncidentType.hackerBreach]!.sequenceLength,
        random,
      );
      return ActiveIncident(
        type: type,
        startedAt: now,
        expiresAt: expiresAt,
        sequence: result['sequence']!,
        buttons: result['buttons']!,
        currentStep: 0,
      );
  }
}

// ─── Multiplier while incident is active ───

double getIncidentMultiplier(ActiveIncident? incident) {
  if (incident == null) return 1;
  if (incident.type == IncidentType.memoryLeak) {
    final config = incidentConfigs[IncidentType.memoryLeak]!;
    final elapsedSec =
        (DateTime.now().millisecondsSinceEpoch - incident.startedAt) / 1000;
    final decayProgress =
        (elapsedSec / config.decayDuration).clamp(0.0, 1.0);
    final value = 1 - decayProgress * 0.5;
    return max(config.minMultiplier, value);
  }
  return incidentConfigs[incident.type]!.activeMultiplier;
}

// ─── Time remaining ───

double getIncidentTimeRemaining(ActiveIncident? incident) {
  if (incident == null) return 0;
  return max(
    0,
    (incident.expiresAt - DateTime.now().millisecondsSinceEpoch) / 1000,
  ).toDouble();
}
